fplll/pruner/pruner.h

```diff
@@ -290,47 +390,207 @@ template <class FT> class Pruner
+     not increase significantly.
+  */
+  void optimize_coefficients_tune_prob(/*io*/ vector<double> &pr);
+
+  /**
+     @brief tune the pruning parameter to make the curve a more smooth.
+
+     Optimization process to the pruning parameters to make the curve
+     more smooth if the input has a large gap between consecutive
+     pruning parameters.
+  */
+  void optimize_coefficients_smooth(/*io*/ vector<double> &pr);
+
+  /**
+     @brief main interface to optimize the overall enumeraiton time
+
+     Main interface to optimize the overall enumeraiton time where the
+     target unciton is: single_enumeration_cost divided by succ. probability
+     or expected solutions.
+  */
+  void optimize_coefficients_cost(/*io*/ vector<double> &pr);
+
+  /**
+     @brief main interface to optimize the single enumeraiton time fixing succ. prob.
+
+     Main interface to optimize the single enumeration time with the constraint
+     that the succ. prob (or expected solutions) is fixed.
+  */
+  void optimize_coefficients_prob(/*io*/ vector<double> &pr);
+
+  /**
+     @brief auxiliary function in optimizing the single enumeraiton time
+     fixing the succ. prob.
+
+     Auxiliary function to optimize the single enumeration time with the
+     constraint that the succ. prob (or expected solutions) is fixed. It
+     is used if the given targeted probaility is larger than the one
+     computed using pruning parameters. Then one increases the succ.
+     probability by increasing the pruning parameters. Note since the
+     radius is fixed as input, it may not be possible to achieve the
+     given targeted probability even if all coefficients are (1, ..., 1).
+     In such case, it will try to increase the succ. prob as much as
+     it can and possibly return (1, ..., 1).
+  */
+  void optimize_coefficients_prob_incr(/*io*/ vector<double> &pr);
+
+  /**
+     @brief auxiliary function in optimizing the single enumeraiton time
+     fixing the succ. prob.
+
+     Auxiliary function to optimize the single enumeration time with the
+     constraint that the succ. prob (or expected solutions) is fixed. It
+     is used if the given targeted probaility is smaller than the one
+     computed using pruning parameters. Then one decreases the succ.
+     probability by descreasing the pruning parameters.
+  */
+  void optimize_coefficients_prob_decr(/*io*/ vector<double> &pr);
+
+  /**
+     @brief auxiliary function in optimizing the single enumeraiton time fixing succ. prob.
 
-  /** @brief Compute the cost of a single enumeration */
+     Heuristic tuning procedure which seems to be useful.
+  */
+  void optimize_coefficients_prob_tune(/*io*/ vector<double> &pr);
+
+  /**
+     @brief Main interface to optimize pruning coefficients
+
+     Main interface to optimize pruning coefficients. It will invoke either:
+     optimize_coefficients_cost() or
+     optimize_coefficients_prob()
+     depending on the input "target". If the target is negative (e.g. -1),
```

Comment:
Why overload target like that, isn't this better handled by a flag?

Reply:
Do you mean adding a flag "optimize_with_fixed_prob"  It will still need user to input a target prob. Seems a bit redundant for the extra flag?